edx_notifications/base_data.py

```diff
@@ -0,0 +1,287 @@
+"""
+Base objects that data.py uses
+"""
+
+import json
+import copy
+import inspect
+import dateutil.parser
+
+from datetime import datetime
+
+from weakref import WeakKeyDictionary
+
+
+class TypedField(object):
+    """
+    Field Decscriptors used to enforce correct typing
+    """
+
+    _expected_type = None
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initializer which takes in the type this field
+        should be set it is set
+        """
+
+        self._expected_type = [args[0]] if not isinstance(args[0], list) else args[0]
+        self._default = kwargs.get('default', None)
+        self._data = WeakKeyDictionary()
+
+    def __get__(self, instance, owner):
+        """
+        Descriptor getter
+        """
+        if not instance:
+            # called as a class method, so return the descriptor (ourselves)
+            return self
+
+        return self._data.get(instance, self._default)
+
+    def __set__(self, instance, value):
+        """
+        Descriptor setter. Be sure to enforce type on setting. But None is allowed.
+        """
+
+        value_type = type(value)
+
+        if value and type(value) not in self._expected_type:
```

Comment:
Use `isinstance`, rather than `type(value) in self._expected_type`, otherwise you won't handle type subclasses.